### PR TITLE
[ENG-7968][eas-cli][eas-json] print deprecation warnings for deprecated `eas.json` fields

### PR DESCRIPTION
**Pull request by szdziedzic**
# Checklist

- [ ] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

https://linear.app/expo/issue/ENG-7968/do-not-cache-podfilelock-by-default

Companion to https://github.com/expo/turtle-v2/pull/1256.

# How

Add a method in `eas.json` which returns deprecation warnings. Print them in the EAS CLI.

# Test Plan

Test manually
Screenshot 2023-04-06 at 16 35 39

### CODE REVIEW

File: packages/eas-json/src/utils.ts [hunk 1]
```diff
@@ -23,6 +23,26 @@ export class EasJsonUtils {
     return resolveBuildProfile({ easJson, platform, profileName });
   }
 
+  public static getBuildProfileDepreactionWarnings(
```
Comment (dsokal) [inline packages/eas-json/src/utils.ts:26]:
```suggestion
  public static getBuildProfileDeprecationWarnings(
```

File: packages/eas-json/src/utils.ts [hunk 2]
```diff
@@ -23,6 +23,26 @@ export class EasJsonUtils {
...
+    buildProfile: BuildProfile,
+    profileName?: string
+  ): { message: string[]; docsUrl?: string }[] {
+    const warnings: { message: string[]; docsUrl?: string }[] = [];
```
Comment (dsokal) [inline packages/eas-json/src/utils.ts:30]: IMO it deserves a separate type
```suggestion
  ): Xyz[] {
    const warnings: Xyz[] = [];
```

File: packages/eas-cli/src/utils/profiles.ts [hunk 3]
```diff
@@ -70,3 +76,25 @@ async function readProfileAsync<T extends ProfileType>({
...
+
+function maybePrintBuildProfileDeprecationWarnings(
+  buildProfile: BuildProfile<Platform>,
+  profileName?: string
+): void {
+  const deprecationWarnings = EasJsonUtils.getBuildProfileDepreactionWarnings(
+    buildProfile,
+    profileName
+  );
+  if (deprecationWarnings.length > 0) {
```
Comment (dsokal) [inline packages/eas-cli/src/utils/profiles.ts:88]: This saves one level of indentation. 
```suggestion
  if (deprecationWarnings.length === 0) {
    return;
  }
```

File: packages/eas-cli/src/utils/profiles.ts [hunk 4]
```diff
@@ -57,11 +59,15 @@ async function readProfileAsync<T extends ProfileType>({
...
   if (type === 'build') {
-    return (await EasJsonUtils.getBuildProfileAsync(
+    const buildProfile = await EasJsonUtils.getBuildProfileAsync(
       easJsonAccessor,
       platform,
       profileName
-    )) as EasProfile<T>;
+    );
+
+    maybePrintBuildProfileDeprecationWarnings(buildProfile, profileName);
```
Comment (dsokal) [inline packages/eas-cli/src/utils/profiles.ts:68]: Maybe we should somehow check if this was already called. So we don't print the same warning multiple times when calling `readProfileAsync`.

**Review by dsokal — APPROVED** [review]

**Review by wkozyra95 — APPROVED** [review]

**Comment by szdziedzic** [comment]
/changelog-entry chore Print deprecation warnings for deprecated `eas.json` fields

**Pull request merged**